core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -0,0 +1,643 @@
+		if (hashdata->ctx.length) {
+			HASH_TRACE("Update Operation");
+			/* Algo Operation - Update */
+			caam_desc_add_word(desc, HASH_UPDATE(alg->type));
+			/* Running context to restore */
+			caam_desc_add_word(desc,
+					   LD_NOIMM(CLASS_2, REG_CTX,
+						    hashdata->ctx.length));
+			caam_desc_add_ptr(desc, hashdata->ctx.paddr);
+		} else {
+			HASH_TRACE("Init Operation");
+
+			/* Algo Operation - Init */
+			caam_desc_add_word(desc, HASH_INIT(alg->type));
+
+			hashdata->ctx.length = alg->size_ctx;
+		}
+
+		if (hashdata->blockbuf.filled != 0) {
+			/* Add the temporary buffer */
+			caam_desc_add_word(desc,
+					   FIFO_LD_EXT(CLASS_2, MSG, NOACTION));
+			caam_desc_add_ptr(desc, hashdata->blockbuf.buf.paddr);
+			caam_desc_add_word(desc, hashdata->blockbuf.filled);
+
+			/* Clean the circular buffer data to be loaded */
+			cache_operation(TEE_CACHECLEAN,
+					hashdata->blockbuf.buf.data,
+					hashdata->blockbuf.filled);
+			hashdata->blockbuf.filled = 0;
+		}
+
+		/* Add the input data multiple of blocksize */
+		caam_desc_add_word(desc, FIFO_LD_EXT(CLASS_2, MSG, LAST_C2));
+		caam_desc_add_ptr(desc, paddr_data);
+		caam_desc_add_word(desc, size_inmade);
+
+		/* Clean the input data to be loaded */
+		cache_operation(TEE_CACHECLEAN, (void *)data, size_inmade);
+
+		/* Save the running context */
+		caam_desc_add_word(desc, ST_NOIMM(CLASS_2, REG_CTX,
+						  hashdata->ctx.length));
+		caam_desc_add_ptr(desc, hashdata->ctx.paddr);
+
+		HASH_DUMPDESC(desc);
+
+		/* Invalidate Context register */
+		cache_operation(TEE_CACHEINVALIDATE, hashdata->ctx.data,
+				hashdata->ctx.length);
+
+		jobctx.desc = desc;
+		retstatus = caam_jr_enqueue(&jobctx, NULL);
+
+		if (retstatus == CAAM_NO_ERROR) {
+			ret = TEE_SUCCESS;
+			HASH_DUMPBUF("CTX", hashdata->ctx.data,
+				     hashdata->ctx.length);
+		} else {
+			HASH_TRACE("CAAM Status 0x%08" PRIx32, jobctx.status);
+			ret = job_status_to_tee_result(jobctx.status);
+		}
+	} else {
+		ret = TEE_SUCCESS;
+
+		if (size_topost) {
+			/* All input data must be saved */
+			size_inmade = 0;
+		}
+	}
+
+	if ((size_topost) && (data)) {
+		struct caambuf indata = { .data = (uint8_t *)data,
+					  .length = inlen };
+
+		HASH_TRACE("Post %zu of input len %zu made %zu", size_topost,
+			   len, size_inmade);
+		ret = caam_cpy_block_src(&hashdata->blockbuf, &indata,
+					 size_inmade);
+	}
+
+exit_update:
+	if (ret != TEE_SUCCESS)
+		do_free_intern(hashdata);
+
+	return ret;
+}
+
+/*
+ * Finalize the Hash operation
+ *
+ * @ctx     Operation Software context
+ * @len     Digest buffer length
+ * @digest  [out] Hash digest buffer
+ */
+static TEE_Result do_final(struct crypto_hash_ctx *ctx, uint8_t *digest,
+			   size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+	struct hashdata *hashdata = hash->ctx;
+	const struct hashalg *alg = NULL;
+
+	struct caam_jobctx jobctx = { 0 };
```

Comment:
Structs are normally initialized with `{}`;

Reply:
Yes but as already say, it doesn't change anything result is the same. Reading `{ 0 }` for me is clearer.

Reply:
I was commenting since we had a discussion some time ago where we agreed on how to initialize the different kinds of variables. We are trying to keep the code as homogenous as possible.

Reply:
Ok. I'll do the change of all structures initialization to remove the `0`. I'll recommand you to add it as a rule in the coding standards because this is not a mandatory rule today.

Reply:
`{ }` is mentioned here: https://optee.readthedocs.io/general/coding_standards.html.
That being said, it is true that optee_os has many occurrences of `{ 0 }`. Perhaps we should replace them.

Reply:
Yes this is mention but for my point of view this is no clearly say that it a rule to follow. We can read it as a suggestion.
I would advice you to add a specific MUST paragraph and SHOULD if this is a recommandation.

Reply:
The only MUST we have is that stuff works and that nothing that needs to be written in coding guidelines. The rest is SHOULD or we wouldn't have written it at all. RFC 2119 puts this quite well as:
_This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course._